File: src/java/magma/app/compile/rule/OrState.java
package magma.app.compile.rule;

import jvm.api.collect.Lists;
import magma.api.collect.List_;
import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.api.result.Err;
import magma.api.result.Ok;
import magma.api.result.Result;
import magma.app.compile.CompileError;

public record OrState<T>(Option<T> maybeValue, List_<CompileError> errors) {
    public OrState() {
        this(new None<>(), Lists.empty());
    }

    public OrState<T> withValue(T result) {
        return new OrState<>(new Some<>(result), errors);
    }

    public boolean isPresent() {
        return maybeValue.isPresent();
    }

    public OrState<T> withErr(CompileError error) {
        return new OrState<T>(maybeValue, errors.add(error));
    }

    public Result<T, List_<CompileError>> toResult() {
        return maybeValue.<Result<T, List_<CompileError>>>map(Ok::new).orElseGet(() -> new Err<>(errors));
    }
}

File: src/clang/magma/compile/rule/OrState.c
package magma.compile.rule;class package magma.compile.rule;{package magma.compile.rule;

import magma.option.None;class import magma.option.None;{

import magma.option.None;
import magma.option.Option;class import magma.option.Option;{
import magma.option.Option;
import magma.option.Some;class import magma.option.Some;{
import magma.option.Some;

public record OrState(Option<String> result) {
    public OrState() {
        this(new None<>());
    }

    public OrState withValue(String value) {
        return new OrState(new Some<>(value));
    }

    public Option<String> toOption() {
        return result;
    }
}
class public record OrState(Option<String> result) {
    public OrState() {
        this(new None<>());
    }

    public OrState withValue(String value) {
        return new OrState(new Some<>(value));
    }

    public Option<String> toOption() {
        return result;
    }
}{

public record OrState(Option<String> result) {
    public OrState() {
        this(new None<>());
    }

    public OrState withValue(String value) {
        return new OrState(new Some<>(value));
    }

    public Option<String> toOption() {
        return result;
    }
}
